ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -0,0 +1,154 @@
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import io.confluent.ksql.function.udf.PluggableUdf;
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.SchemaUtil;
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
+
+public class UdfLoader {
+
+  private static final Logger logger = LoggerFactory.getLogger(UdfLoader.class);
+
+  private final MetaStore metaStore;
+  private final File pluginDir;
+  private final ClassLoader parentClassLoader;
+  private final Predicate<String> blacklist;
+  private final UdfCompiler compiler;
+  private final boolean loadCustomerUdfs;
+
+
+  public UdfLoader(final MetaStore metaStore,
+                   final File pluginDir,
+                   final ClassLoader parentClassLoader,
+                   final Predicate<String> blacklist,
+                   final UdfCompiler compiler,
+                   final boolean loadCustomerUdfs) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore can't be null");
+    this.pluginDir = Objects.requireNonNull(pluginDir, "pluginDir can't be null");
+    this.parentClassLoader = Objects.requireNonNull(parentClassLoader,
+        "parentClassLoader can't be null");
+    this.blacklist = Objects.requireNonNull(blacklist, "blacklist can't be null");
+    this.compiler = Objects.requireNonNull(compiler, "compiler can't be null");
+    this.loadCustomerUdfs = loadCustomerUdfs;
+  }
+
+  public void load() {
+    // load udfs packaged as part of ksql first
+    loadUdfs(parentClassLoader);
+    if (loadCustomerUdfs) {
+      try {
+        Files.find(pluginDir.toPath(), 1, (path, attributes) -> path.toString().endsWith(".jar"))
+            .map(path -> UdfClassLoader.newClassLoader(path, parentClassLoader, blacklist))
+            .forEach(this::loadUdfs);
+      } catch (IOException e) {
+        logger.error("Failed to load UDFs from location {}", pluginDir, e);
+      }
+    }
+  }
+
+
+  private void loadUdfs(final ClassLoader loader) {
+    new FastClasspathScanner()
+        .overrideClassLoaders(loader)
+        .ignoreParentClassLoaders()
+        // if we are loading from the parent classloader then restrict the name space to only
+        // jars/dirs containing "ksql-engine". This is so we don't end up scanning every jar
+        .filterClasspathElements(
+            name -> parentClassLoader != loader || name.contains("ksql-engine"))
+        .matchClassesWithMethodAnnotation(Udf.class,
+            (theClass, executable) -> {
+              final UdfDescription annotation = theClass.getAnnotation(UdfDescription.class);
+              if (annotation != null) {
+                final Method method = (Method) executable;
```

Comment:
Is there any way this won't be a Method? e.g. misplaced annotation?

Reply:
annotation is a Method level annotation so their code wont compile

Reply:
Cool. Just checking.